### DISCUSSION TITLE
"Illegal instruction (core dumped)" when trying to run ckb from linux terminal

### DISCUSSION THREAD

**Discussion by ItalianNoob** [nervosnetwork/ckb#2654] [General]
**Environment**

- **CKB version**: none
- **Chain**: none
- **Operating system**: Linux 
- **Arch**: x64
- **Installation**: GitHub Release 

**Describe the bug**

I tried to install ckb from the release file on github version 0.40 but whenever I try to run it I get the "illegal instruction (core dumped) error.

**Comment by driftluo** [comment]
Illegal instruction means that the CPU of your machine does not support some common instructions of modern CPU, such as `+aes,+pclmul,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2`

Of course, the GitHub release version is not very radically optimized. It only uses the part of the machine support provided by Travis. In theory, it should only be `sse1,sse2,cmov,mode64` 

If you really want to run ckb on this machine, it is recommended to compile ckb binary from the source code on this machine, and then use.

Before compiling, you most likely need to install some development environments, there may be:
```
$ apt install libssl-dev pkg-config autoconf libtool git
$ curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```
then:
```
$ git clone https://github.com/nervosnetwork/ckb.git
$ cd ckb
$ make prod
```

**Comment by gavriilos** [comment]
Building locally didn't fix this for me. I am building the latest release (`0.104.0`).

I tried setting the env vars `RUSTFLAGS` and `RUSTCFLAGS` to `-C target-cpu=native` but I still get an illegal instruction.
Some component is getting built for a target that is not my local CPU.

I had no such problem with previous versions. Version `0.103.0` works fine here.

Any idea of what component may be the culprit? Or how I may go about debugging this?

**Reply by driftluo** [reply]
May you should use `make prod_portable` to build binary

The reason for this is that rocksdb enables the optimization of sse2 and other instructions by default, and the portable feature must be enabled to turn off these optimizations.

**Reply by gavriilos** [reply]
Thanks a lot for the quick help @driftluo !

(I guess halfway there now.)

`make prod_portable` builds out a working executable. Unfortunately I had to run it on the `rc/v0.104.0.x` branch.
On both `v0.104.0` and `master` `make prod_portable` fails with the following error:

    RUSTFLAGS="${RUSTFLAGS} --cfg disable_faketime" cargo build  --profile prod --features "with_sentry,with_dns_seeding,ckb-db/portable"
    error: none of the selected packages contains these features: ckb-db/portable, did you mean: ckb/portable?
    make: *** [Makefile:135: prod_portable] Error 101

With `rc/v0.104.0.x` there are no illegal instructions but the ckb process crashes on start with the following error:

    Started Nervos CKB Node.
    2022-09-02 16:29:02.651 +00:00 main ERROR panic  thread 'main' panicked at 'Start network service failed: P2P(Transport(Io(Os { code: 22, kind: InvalidInput, message: "Invalid argument"  })))': util/launcher/src/lib.rs:317   0: ckb_logger_service::setup_panic_logger::{{closure}}
       1: std::panicking::rust_panic_with_hook
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:702:17
       2: std::panicking::begin_panic_handler::{{closure}}
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:588:13
       3: std::sys_common::backtrace::__rust_end_short_backtrace
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:138:18
       4: rust_begin_unwind
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
       5: core::panicking::panic_fmt
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
       6: core::result::unwrap_failed
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:1785:5
       7: ckb_launcher::Launcher::start_network_and_rpc
       8: ckb_bin::subcommand::run::run
       9: ckb_bin::run_app
      10: ckb::main
      11: std::sys_common::backtrace::__rust_begin_short_backtrace
      12: std::rt::lang_start::{{closure}}
      13: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:259:13
          std::panicking::try::do_call
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
          std::panicking::try
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
          std::panic::catch_unwind
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
          std::rt::lang_start_internal::{{closure}}
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:48
          std::panicking::try::do_call
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
          std::panicking::try
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
          std::panic::catch_unwind
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
          std::rt::lang_start_internal
                 at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:20
      14: main
      15: __libc_start_main
                 at ./csu/../csu/libc-start.c:308:16
      16: _start

Any suggestion of which branch I should try? I wonder if the code at `rc/v0.104.0.x` is ready for mainnet usage.

**Reply by driftluo** [reply]
I tried to compile and start with `make prod_portable` with `rc/v0.104.0.x` branch and it worked fine

It seems that your crash is because the configuration file is wrong

**Reply by gavriilos** [reply]
Indeed! The issue was my config file. I had it configured like this to accept IPv6 connections:

    [network]
    listen_addresses = ["/ip4/0.0.0.0/tcp/8115", "/ip6/::/tcp/8115"]

This works out fine on `v0.103.0` but fails on `rc/v0.104.0.x`. I changed it to:

    [network]
    listen_addresses = ["/ip6/::/tcp/8115"]

Which works on `rc/v0.104.0.x` and strangely accepts both IPv4 and IPv6 connections.

Thanks a bunch for the help @driftluo

**Reply by driftluo** [reply]
https://github.com/nervosnetwork/ckb/pull/3566 Inconsistent behavior will be fixed in this PR

**Reply by gavriilos** [reply]
Wow! Thanks a lot for the quick fix. And thanks for the portable binaries in the latest release: quite handy!
